contracts/contracts/opl/Endpoint.sol

```diff
@@ -0,0 +1,239 @@
+        uint256 fee = estimateFee(_message);
+        ICelerMessageBus(bus).sendMessage{value: fee}(
+            host,
+            hostChainId,
+            abi.encodeWithSelector(bytes4(keccak256(_method)), txSeq, _message)
+        );
+        ++txSeq;
+    }
+
+    function executeMessage(
+        address _sender,
+        uint64 _senderChainId,
+        bytes calldata _message,
+        address
+    ) external payable returns (uint256) {
+        if (msg.sender != bus) revert Forbidden("bridge endpoint");
+        if (_sender != host || _senderChainId != hostChainId)
+            revert Forbidden("non-host sender");
+        bytes4 epSel = bytes4(_message[:4]);
+        uint256 seq = uint256(bytes32(_message[4:36]));
+        bytes calldata message = _message[36:];
+        if (inOrder) {
+            if (seq != rxSeq) revert WrongSeqNum(rxSeq, seq);
+            ++rxSeq;
+        }
+        function(bytes calldata) returns (Result) ep = endpoints[epSel];
+        bool epExists;
+        assembly ("memory-safe") {
+            epExists := not(iszero(ep))
+        }
+        Result result = Result.PermanentFailure;
+        if (epExists) result = endpoints[epSel](message);
+        // Convert the Result to a Celer ExecutionStatus.
+        if (result == Result.TransientFailure) return 2; // ExecutionStatus.Retry
+        if (result == Result.Success) return 1; // ExecutionStatus.Success
+        return 0; // ExecutionStatus.Fail
+    }
+
+    function estimateFee(bytes memory _message)
+        internal
+        view
+        returns (uint256)
+    {
+        uint256 feeBase = ICelerMessageBus(bus).feeBase();
+        uint256 feePerByte = ICelerMessageBus(bus).feePerByte();
+        return
+            feeBase +
+            (_message.length +
+                32 + /* seq */
+                4) * /* epsel */
+            feePerByte;
+    }
+
+    function _chainName2ChainId(bytes32 name) internal pure returns (uint256) {
+        if (name == "ethereum") return 1;
+        if (name == "goerli") return 5;
+        if (name == "optimism") return 10;
+        if (name == "bsc") return 56;
+        if (name == "bsc-testnet") return 97;
+        if (name == "polygon") return 137;
+        if (name == "fantom") return 0xfa;
+        if (name == "fantom-testnet") return 0xfa2;
+        if (name == "moonriver") return 0x505;
+        if (name == "sapphire-testnet") return 0x5aff;
+        if (name == "arbitrum-one") return 1;
+        if (name == "arbitrum-nova") return 1;
+        if (name == "sapphire") return 0x5afe;
+        if (name == "polygon-mumbai") return 80001;
+        if (name == "avalanche") return 43114;
+        if (name == "avalanche-fuji") return 43313;
+        if (name == "arbitrum-testnet") return 0x66eeb;
+        return 0;
+    }
+
+    function _getChainConfig(uint256 _chainId)
+        internal
+        pure
+        returns (address _bus, bool _isTestnet)
+    {
+        if (_chainId == 1)
```

Comment:
Would be nice to put the source for these in a comment (I assume these are from https://im-docs.celer.network/developer/contract-addresses-and-rpc-info).